### PR TITLE
Browser integration

### PR DESCRIPTION
**Pull request by moliholy**
After making the browser stable here I have a few modifications for the library.
- Add considerably more tag support for both the History and Repository.
- Possibility to change the repository revision (consequently changing the root catalog). By default the first root catalog will be the one indicated in the manifest
- Force the cache creation in a directory that already exists so that the user doesn't run out of file descriptors naïvely creating repositories
- Utilities for dirent

### CODE REVIEW

File: cvmfs/repository.py [hunk 1]
```diff
@@ -145,10 +145,11 @@ def __init__(self, fetcher):
         self._read_manifest()
         self._try_to_get_last_replication_timestamp()
         self._try_to_get_replication_state()
+        self._root_catalog_hash = self.manifest.root_catalog
 
     @classmethod
     def from_source(cls, source, cache_dir = None):
-        if source == '':
+        if not source:
```
Comment (moliholy) [inline cvmfs/repository.py:152]: This checks if the source is either empty or None

File: cvmfs/dirent.py [hunk 2]
```diff
@@ -36,6 +36,17 @@ def to_suffix(content_hash_type):
         else:
             return ""
 
+    @staticmethod
+    def to_string(hash_type):
```
Comment (reneme) [inline cvmfs/dirent.py:40]: Where is that used?

File: cvmfs/dirent.py [hunk 3]
```diff
@@ -36,6 +36,17 @@ def to_suffix(content_hash_type):
...
+        if hash_type == -1:
+            return 'Unknown'
+        elif hash_type == 1:
+            return 'Sha1'
+        elif hash_type == 2:
+            return 'Ripemd160'
+        elif hash_type == 3:
+            return 'UpperBound'
```
Comment (reneme) [inline cvmfs/dirent.py:48]: What is `UpperBound`?

File: cvmfs/history.py [hunk 4]
```diff
@@ -12,9 +12,25 @@
 class RevisionTag:
     """ Specific revisions in CernVM-FS 2.1.x repositories have named tags  """
     @staticmethod
-    def sql_query():
-        return '''SELECT name, hash, revision, timestamp, channel, description
-                  FROM tags ORDER BY timestamp DESC'''
+    def sql_query_all():
+        return 'SELECT name, hash, revision, timestamp, channel, description \
```
Comment (reneme) [inline cvmfs/history.py]: please DRY this up a little. Like so maybe:

``` python
return "SELECT" + _database_fields() + " FROM tags ..."
```

File: cvmfs/history.py [hunk 5]
```diff
@@ -54,9 +70,24 @@ def __iter__(self):
...
+    def get_tag_by_revision(self, revision):
+        result = self.run_sql(RevisionTag.sql_query_revision(revision))
+        if result:
+            return RevisionTag(result[0])
+
+    def get_tag_by_date(self, timestamp):
+        result = self.run_sql(RevisionTag.sql_query_date(timestamp))
+        if result:
+            return RevisionTag(result[0])
+
```
Comment (reneme) [inline cvmfs/history.py:94]: Let's DRY those up a bit like this:

``` python
def _get_tag_by_query(self, query):
  result = self.run_sql(query)
  if result:
    return RevisionTag(result[0])
```

The others should then just call into `_get_tag_by_query()` with the appropriate query string.

File: cvmfs/dirent.py (hunk 2 above)
Comment (moliholy) [inline cvmfs/dirent.py:40]: In the browser. It displays the hash and its type

File: cvmfs/dirent.py (hunk 3 above)
Comment (moliholy) [inline cvmfs/dirent.py:48]: No idea, it is one of the options! Look a few lines on the top

**Comment by reneme** [comment]
One general remark:
I think the `Repository` class now finally gets out of hand by being able to switch between revisions. That makes it actually quite stateful in a sense that the following could become confusing:

``` python
import cvmfs
repo = cvmfs.open_repository("http://cvmfs-stratum-one.cern.ch/cvmfs/atlas.cern.ch")
repo.switch_revision(1337) # very old revision
# 200 more python lines here
dirent = repo.lookup_path("/a/funny/cutting/edge/entry")
# dirent not found... wut?
```

I sense that it would be better to have a class encapsulating a single catalog tree revision like:

``` python
repo = ...
catalogs_1337 = repo.open_revision(1337)
catalogs_yesterday = repo.open_tag("yesterday")
catalogs_cutting_edge = repo.open_latest() # based on .cvmfspublished root hash
catalogs_***.lookup("/a/funny/cutting/edge/entry")
```

Hence `Repository::open_***()` would return a `CatalogManager` that is based on a specific root hash which cannot be altered. Opening a new revisions means instantiating a new `CatalogManager` in that model. Of course there is one major caveat: Catalogs might show up in multiple `CatalogManagers`.

This is just jotting down thoughts. Let's think about how to do that in a concise way.

**Comment by moliholy** [comment]
In that case, shall this PR be approved and those changes you are mentioning added to a new one?
I have already made the rest of the changes.

**Comment by reneme** [comment]
Let's do it in a separate Pull Request to keep things simple.

**Pull request merged**